[KEP-4817]: DRA: Resource Claim Status with possible standardized network interface data (#4861)
kubernetes/enhancements: keps/sig-node/4817-resource-claim-device-status/README.md

## Motivation

As of now, when a device is configured in a pod/container, the state and characteristics of the device set during the configuration stage are invisible. The lack of this information is then a challenge for the users and developers to diagnose issues, verify configurations, and integrate the allocated resources into higher-level services. Reporting this information is then crucial in environments where device specific configurations are necessary.

For certain types of devices, such as network interfaces, knowing the detailed characteristics of what has been allocated is particularly useful and crucial. For example, reporting the interface Name, MAC address and IP addresses of network interfaces in the status of a ResourceClaim can significantly help in configuring and managing network services, as well as in debugging network related issues. By including such device specific information, this proposal addresses existing gaps in visibility and facilitates better integration and management of resources.

### Goals

* Allow arbitrary, driver-specific information to be reported from the DRA drivers for each allocated device in a ResourceClaim.

> * Establish a foundation for potential future standardization (e.g. Network Devices status).

Author: I fixed it, thank you. Yes, I tried to follow option 5.

* Enable 3rd party implementation of new functionalities based on the Device Status (e.g. Secondary Network Services if the IPs of a network device are reported).

### Non-Goals

* Implement new functionalities based on the Device Status (e.g. Kubernetes EndpointSlice Controller supporting IPs reported by networking DRA drivers).
* Modifying the Resource Claim workflow to include the device status report.

## Proposal

### API - ResourceClaim.Status

The API changes define a new `DeviceStatuses` field in the existing `ResourceClaimStatus` struct. `DeviceStatuses` is a slice of a new struct `AllocatedDeviceStatus` which holds device specific information. 

> A device, identified by `<driver name>/<pool name>/<device name>` can be represented only once in the `DeviceStatuses` slice and will also mention which request caused the device to be allocated. The state and characteristics of the device are reported in the `Conditions`, representing the operational state of the device and in the `DeviceInfo`, an arbitrary data slice representing device specific characteristics. Additionally, for networking devices, a field `NetworkDeviceInfo` can be used to report the IPs, the MAC address and the interface name.

Reviewer: This is one approach, where there is `DeviceInfo` and `NetworkDeviceInfo`. I am not sure that will pass API review muster, but maybe. An alternative is what is described in Option 5. Instead of 3 fields, there would be two fields: `Conditions` and `DeviceInfo`. The `DeviceInfo` field is a one-of API type. So, something like:

```go
type AllocatedDeviceStatus struct {
  ...
  DeviceInfo *DeviceInfo
}

type DeviceInfo struct {
    // +optional
	// +oneOf=DeviceInfoType
    Network *DeviceInfoNetwork

   // +optional
   // +oneOf=DeviceInfoType
   General *DeviceInfoGeneral
}

type DeviceInfoGeneral struct {
  Opaque RawExtension
}

type DeviceInfoNetwork struct {
   InterfaceName string
   IPAddresses []IPAddress
   MACAddress string

  Opaque RawExtension
}
```

@thockin WDYT? *Every* one-of would have the `Opaque RawExtension` plus one-of specific fields. Should that therefore just bubble up one level, so that we have a RawExtension, and then an optional device-type based one-of? Or, what about compound devices that may be several types. Should we just make the `Opaque` and `Network` fields separate, optional fields that may or may not be populated for any given device? I am not sure the one-of construction actually helps. (In other words, go with what is already here in the KEP).

Author: I see, I interpreted it like any driver can use the opaque field (`DeviceInfo`) + their own type. So the `oneOf` would be between specific type (`NetworkDeviceInfo`, `GPUDeviceInfo`...).

Reviewer: Yeah, but the way you did it is less prescriptive - it allows a device to serve multiple functions. So I rather like that.

Reviewer: To be a little more verbose about my last comment:

1) we don't know what devices will come out in the future. So, whose to say they won't stick an optical connection on a GPU...or maybe they have already? So, maybe the GPU and NIC will be a single device, not two separate devices

2) As VM architects, there may be specific devices that "go together as one". The GPU/NIC combos is an example. Right now, we are using MatchAttributes with a pcieRoot attribute to align the internal topology. That's a lot of work for the user. Instead, I am looking at building a "compound device driver" on top of the partitionable device model. This would allow a single "device" owned by that driver to be allocated, but in reality that driver would proxy calls for that device to multiple other drivers. So, the VM architect could "pair" specific devices, such that you cannot get one without the other. Those devices would have separate functions, represented by the separate underlying drivers. I would want to be able to stick information about each underlying device in the right "device info field" for that function.

Thinking about it, this is still limited of course, to one device per function. Another option would be to have the one-of, but allow multiple entries per driver/pool/device tuple.

	// needed on a node.
	//
	// Must be a DNS subdomain and should end with a DNS domain owned by the
	// vendor of the driver.
	//
	// +required
	Driver string `json:"driver" protobuf:"bytes,2,rep,name=driver"`

	// This name together with the driver name and the device name field
	// identify which device was allocated (`<driver name>/<pool name>/<device name>`).
	//
	// Must not be longer than 253 characters and may contain one or more
	// DNS sub-domains separated by slashes.
	//
	// +required
	Pool string `json:"pool" protobuf:"bytes,3,rep,name=pool"`

	// Device references one device instance via its name in the driver's
	// resource pool. It must be a DNS label.
	//
	// +required
	Device string `json:"device" protobuf:"bytes,4,rep,name=device"`

	// Conditions contains the latest observation of the device's state.
	// If the device has been configured according to the class and claim
	// config references, the `Ready` condition should be True.
	//
	// +optional
	// +listType=atomic
	Conditions []metav1.Condition `json:"conditions" protobuf:"bytes,5,rep,name=conditions"`

	// DeviceInfo contains Arbitrary driver-specific data.
	//
	// +optional
	// +listType=atomic

> 	DeviceInfo []runtime.RawExtension `json:"deviceInfo,omitempty" protobuf:"bytes,6,rep,name=deviceInfo"`

Reviewer: see comment above regarding this versus a one-of

specific characteristics. Additionally, for networking devices, a field
`NetworkDeviceInfo` can be used to report the IPs, the MAC address and the
interface name.

`DeviceInfo` being a slice of arbitrary data allows the DRA Driver to store
device specific data in different formats. For example, a Network Device being
configured by via a CNI plugin could get its `DeviceInfo` filled with the CNI
result for troubleshooting purpose and with a Network result in a modern
standard format (closer to Pod.Status.PodIPs for example) used by 3rd party
controllers.

For each device, if required, the DRA Driver processing the device allocation
can report the status of it in the `Status.DeviceStatuses` of the ResourceClaim
by using the Kubernetes API.

```golang
// ResourceClaimStatus tracks whether the resource has been allocated and what
// the result of that was.
type ResourceClaimStatus struct {
	...

	// DeviceStatuses contains the status of each device allocated for this
	// claim, as reported by the driver. This can include driver-specific
	// information. Entries are owned by their respective drivers.
	//
	// +optional
	// +listType=map
	// +listMapKey=devicePoolName

> 	// +listMapKey=deviceName

Reviewer 2: `driver` also needs to be a key.

	DeviceStatuses []AllocatedDeviceStatus `json:"deviceStatuses,omitempty" protobuf:"bytes,4,opt,name=deviceStatuses"`
}

// AllocatedDeviceStatus contains the status of an allocated device, if the
// driver chooses to report it. This may include driver-specific information.
type AllocatedDeviceStatus struct {
	// Request is the name of the request in the claim which caused this
	// device to be allocated. Multiple devices may have been allocated
	// per request.
	//
	// +required

> 	Request string `json:"request" protobuf:"bytes,1,rep,name=request"`

Reviewer 2: Is this needed? More generally, can we drop the "Allocated" part from the API?

It is not useful to post status of arbitrary, unrelated devices, but it can happen at least temporarily, for example directly after a claim got deallocated.

Speaking of which, who removes such stale entries?

A claim can go from "new" to "allocated" back to "deallocated" and then to "allocated" again.

Author: Right, I removed it, I guess this is repetitive as this information can be retrieved from `.Status.Allocation.Devices.Results`.

In the `Notes/Constraints/Caveats` section, I stated `The accuracy of the information depends on the implementation of the DRA Drivers, the reported status of the device may not always reflect the real time changes of the state of the device.` so the DRA-Driver should update these information based on when it configures/check/unconfigured devices.

Probably if the `ResourceClaim` gets deallocated then the `DeviceStatuses` can be cleared as it might be deleted or reallocated in which case the driver will have to reconfigure the devices in the pods/containers?

Reviewer 2: I suppose that could be added. In that case, calling this `AllocatedDeviceStatus` would make sense because it is meant and documented as "only for allocated devices".

	// plugin should be invoked to process the allocation once the claim is
	// needed on a node.
	//
	// Must be a DNS subdomain and should end with a DNS domain owned by the
	// vendor of the driver.
	//
	// +required
	Driver string `json:"driver" protobuf:"bytes,2,rep,name=driver"`

	// This name together with the driver name and the device name field
	// identify which device was allocated (`<driver name>/<pool name>/<device name>`).
	//
	// Must not be longer than 253 characters and may contain one or more
	// DNS sub-domains separated by slashes.
	//
	// +required
	Pool string `json:"pool" protobuf:"bytes,3,rep,name=pool"`

	// Device references one device instance via its name in the driver's
	// resource pool. It must be a DNS label.
	//
	// +required
	Device string `json:"device" protobuf:"bytes,4,rep,name=device"`

	// Conditions contains the latest observation of the device's state.
	// If the device has been configured according to the class and claim
	// config references, the `Ready` condition should be True.
	//
	// +optional
	// +listType=atomic
	Conditions []metav1.Condition `json:"conditions" protobuf:"bytes,5,rep,name=conditions"`

> 	// DeviceInfo contains Arbitrary driver-specific data.

Reviewer 2: ```suggestion
	// DeviceInfo contains arbitrary driver-specific data.
```

Reviewer 2: How can a consumer reliably determine how to parse this content? The existing `driver` field above might not be enough: for example, two different drivers "foo.example.com" and "bar.example.com" might both provide data according to some common specification.

A consumer shouldn't have to keep a database that lists whether a driver uses such a common specification.

Reviewer: > How can a consumer reliably determine how to parse this content? The existing `driver` field above might not be enough: for example, two different drivers "foo.example.com" and "bar.example.com" might both provide data according to some common specification.
> 
> 
> 
> A consumer shouldn't have to keep a database that lists whether a driver uses such a common specification.
> 
> 

One option is to require a GVK in here. But another is just to say - you shouldn't *be* consuming opaque data. If there is a common spec, it should be a built in type.

Author: I think that was the idea yes as it would be quite difficult to cover all use cases with the specs we would provide in this KEP. Letting the driver reporting arbitrary status could still be useful I guess (for troubleshooting for example)?

Reviewer: yes. but no one except the driver should rely on that data

	//
	// +optional
	// +listType=atomic
	DeviceInfo []runtime.RawExtension `json:"deviceInfo,omitempty" protobuf:"bytes,6,rep,name=deviceInfo"`

    // NetworkDeviceInfo contains network-related information specific to the device.
	//
	// +optional
	NetworkDeviceInfo NetworkDeviceInfo `json:"networkDeviceInfo,omitempty" protobuf:"bytes,7,rep,name=networkDeviceInfo"`
}

// NetworkDeviceInfo provides network-related details for the allocated device.
// This information may be filled by drivers or other components to configure
// or identify the device within a network context.
type NetworkDeviceInfo struct {
	// Interface specifies the name of the network interface associated with
	// the allocated device. This might be the name of a physical or virtual
	// network interface.
	//
	// +optional
	Interface string `json:"interface,omitempty" protobuf:"bytes,1,rep,name=interface"`

	// IPs lists the IP addresses assigned to the device's network interface.
	// This can include both IPv4 and IPv6 addresses.
	//
	// +optional

> 	IPs []string `json:"ips,omitempty" protobuf:"bytes,2,rep,name=ips"`

Reviewer 3: Does it make sense to have a type here instead of string? Does mask, family add value? 

`type IP struct {
    address 
    mask
    family
}`

Author: I replaced this field with a `NetworkAddresses` one:
```golang
type NetworkDeviceInfo struct {
...
    // NetworkAddresses lists the network addresses assigned to the device's network interface.
    // This can include both IPv4 and IPv6 addresses.
    //
    // +optional
    NetworkAddresses []NetworkAddress `json:"networkAddresses,omitempty" protobuf:"bytes,2,rep,name=networkAddresses"`
...
}

// NetworkAddress provides a network address related details such as IP and Mask.
type NetworkAddress struct {
    // CIDR contains the network address in CIDR notation, which includes
    // both the address and the associated subnet mask.
    // e.g.: "192.0.2.0/24" for IPv4 and "2001:db8::/64" for IPv6.
    //
    // +required
    CIDR string `json:"cidr,omitempty" protobuf:"bytes,1,rep,name=cidr"`
}
```
For reference here is how the [NetworkAttachmentDefinition Status](https://github.com/k8snetworkplumbingwg/network-attachment-definition-client/blob/v1.7.3/pkg/apis/k8s.cni.cncf.io/v1/types.go#L103) looks like and here is for the [CNI Result](https://github.com/containernetworking/cni/blob/v1.2.3/pkg/types/100/types.go#L89).

Author: I switched it back to a single field
```golang
type NetworkDeviceData struct {
    ...
    // Addresses lists the network addresses assigned to the device's network interface.
    // This can include both IPv4 and IPv6 addresses.
    // The addresses are in the CIDR notation, which includes both the address and the
    // associated subnet mask.
    // e.g.: "192.0.2.0/24" for IPv4 and "2001:db8::/64" for IPv6.
    //
    // +optional
    // +listType=atomic
    Addresses []string `json:"addresses,omitempty" protobuf:"bytes,2,opt,name=addresses"`
    ...
}
```

configuring and managing network services, as well as in debugging network
related issues. By including such device specific information, this proposal
addresses existing gaps in visibility and facilitates better integration and
management of resources.

### Goals

* Allow arbitrary, driver-specific information to be reported from the DRA
  drivers for each allocated device in a ResourceClaim.
* Establish a standardization for device status (e.g. Network Devices status).
* Enable 3rd party implementation of new functionalities based on the Device
  Status (e.g. Secondary Network Services if the IPs of a network device are
reported).

### Non-Goals

* Implement new functionalities based on the Device Status (e.g. Kubernetes
  EndpointSlice Controller supporting IPs reported by networking DRA drivers).
* Modifying the Resource Claim workflow to include the device status report.

## Proposal

### API - ResourceClaim.Status

The API changes define a new `DeviceStatuses` field in the existing
`ResourceClaimStatus` struct. `DeviceStatuses` is a slice of a new struct
`AllocatedDeviceStatus` which holds device specific information. 

> A device, identified by `<driver name>/<pool name>/<device name>` can be

Reviewer 4: @johnbelamaric did we establish this convention on other KEP, or discussed it? I vaguely remember something , maybe in Sergey KEP about the device driver info?

Reviewer 2: Yes, this triple is a core DRA concept. ResourceSlice and allocation result both use it.

represented only once in the `DeviceStatuses` slice and will also mention which
request caused the device to be allocated. The state and characteristics of the
device are reported in the `Conditions`, representing the operational state of
the device and in the `DeviceInfo`, an arbitrary data slice representing device
specific characteristics. Additionally, for networking devices, a field

> `NetworkDeviceInfo` can be used to report the IPs, the MAC address and the

Reviewer 4: I think Device is redundant in the field and `NetworkInfo` is a name that better represents what we want, in my opinion

Author: I renamed it.

interface name.

`DeviceInfo` being a slice of arbitrary data allows the DRA Driver to store
device specific data in different formats. For example, a Network Device being
configured by via a CNI plugin could get its `DeviceInfo` filled with the CNI
result for troubleshooting purpose and with a Network result in a modern
standard format (closer to Pod.Status.PodIPs for example) used by 3rd party
controllers.

For each device, if required, the DRA Driver processing the device allocation
can report the status of it in the `Status.DeviceStatuses` of the ResourceClaim
by using the Kubernetes API.

```golang
// ResourceClaimStatus tracks whether the resource has been allocated and what
// the result of that was.
type ResourceClaimStatus struct {
    ...

    // DeviceStatuses contains the status of each device allocated for this
    // claim, as reported by the driver. This can include driver-specific
    // information. Entries are owned by their respective drivers.
    //
    // +optional
    // +listType=map
    // +listMapKey=driver
    // +listMapKey=device
    // +listMapKey=pool

>     DeviceStatuses []AllocatedDeviceStatus `json:"deviceStatuses,omitempty" protobuf:"bytes,4,opt,name=deviceStatuses"`

Reviewer 4: I personally find it easier to think about if the DeviceStatus is part of the Results, we also have there the driver, device, pool key 

```
ResourceClaimStatus
  Allocation: &AllocationResult{
       Devices: DeviceAllocationResult {
                Results []DeviceRequestAllocationResult
                Config []DeviceAllocationConfiguration 
       NodeSelector *v1.NodeSelector
       Controller  string
  }
```

since this we already have there the Driver, Pool, Device identifier

```go
// AllocationResultsMaxSize represents the maximum number of
// entries in allocation.devices.results.
const AllocationResultsMaxSize = 32

// DeviceRequestAllocationResult contains the allocation result for one request.
type DeviceRequestAllocationResult struct {
	// Request is the name of the request in the claim which caused this
	// device to be allocated. Multiple devices may have been allocated
	// per request.
	//
	// +required
	Request string `json:"request" protobuf:"bytes,1,name=request"`

	// Driver specifies the name of the DRA driver whose kubelet
	// plugin should be invoked to process the allocation once the claim is
	// needed on a node.
	//
	// Must be a DNS subdomain and should end with a DNS domain owned by the
	// vendor of the driver.
	//
	// +required
	Driver string `json:"driver" protobuf:"bytes,2,name=driver"`

	// This name together with the driver name and the device name field
	// identify which device was allocated (`<driver name>/<pool name>/<device name>`).
	//
	// Must not be longer than 253 characters and may contain one or more
	// DNS sub-domains separated by slashes.
	//
	// +required
	Pool string `json:"pool" protobuf:"bytes,3,name=pool"`

	// Device references one device instance via its name in the driver's
	// resource pool. It must be a DNS label.
	//
	// +required
	Device string `json:"device" protobuf:"bytes,4,name=device"`
    // Conditions contains the latest observation of the device's state.
    // If the device has been configured according to the class and claim
    // config references, the `Ready` condition should be True.
    //
    // +optional
    // +listType=atomic
    Conditions []metav1.Condition `json:"conditions" protobuf:"bytes,5,rep,name=conditions"`

    // DeviceInfo contains arbitrary driver-specific data.
    //
    // +optional
    // +listType=atomic
    DeviceInfo []runtime.RawExtension `json:"deviceInfo,omitempty" protobuf:"bytes,6,rep,name=deviceInfo"`

    // NetworkDeviceInfo contains network-related information specific to the device.
    //
    // +optional
    NetworkInfo NetworkInfo `json:"networkInfo,omitempty" protobuf:"bytes,7,rep,name=networkInfo"`
}
```

Reviewer 2: Let's better keep allocation and additional status information separate:
- The `AllocationResult` is currently treated as immutable during validation.
- Different clients are responsible for updating it.

Reviewer 4: ack, clarified offline, the scheduler is also writing the `DeviceRequestAllocationResult`

}

// AllocatedDeviceStatus contains the status of an allocated device, if the
// driver chooses to report it. This may include driver-specific information.

> type AllocatedDeviceStatus struct {

Reviewer 2: See also https://github.com/kubernetes/enhancements/pull/4861#discussion_r1770979535

> type AllocatedDeviceStatus struct {

Reviewer 2: Regarding "not be allocated because it failed or something and we want to report back": if a device cannot be allocated for a claim, then there is no reason to publish anything about it in the claim. Otherwise that same information would have to be added to all claims which potentially might have used such a device.

> type AllocatedDeviceStatus struct {

Reviewer: Seems we are not differentiating between "reserved by the scheduler" and "actuated by the driver" and are using "allocated" for both? We should probably differentiate.

> type AllocatedDeviceStatus struct {

Reviewer 4: To use if I understand it, `AllocatedDeviceStatus` here means "actuated by the driver" , right? since the driver is the one that populates the information ...

> type AllocatedDeviceStatus struct {

Reviewer: Yes.

    // identify which device was allocated (`<driver name>/<pool name>/<device name>`).
    //
    // Must not be longer than 253 characters and may contain one or more
    // DNS sub-domains separated by slashes.
    //
    // +required
    Pool string `json:"pool" protobuf:"bytes,3,rep,name=pool"`

    // Device references one device instance via its name in the driver's
    // resource pool. It must be a DNS label.
    //
    // +required
    Device string `json:"device" protobuf:"bytes,4,rep,name=device"`

    // Conditions contains the latest observation of the device's state.
    // If the device has been configured according to the class and claim
    // config references, the `Ready` condition should be True.
    //
    // +optional
    // +listType=atomic
    Conditions []metav1.Condition `json:"conditions" protobuf:"bytes,5,rep,name=conditions"`

    // DeviceInfo contains arbitrary driver-specific data.
    //
    // +optional
    // +listType=atomic
    DeviceInfo []runtime.RawExtension `json:"deviceInfo,omitempty" protobuf:"bytes,6,rep,name=deviceInfo"`

    // NetworkDeviceInfo contains network-related information specific to the device.
    //
    // +optional

>     NetworkDeviceInfo NetworkDeviceInfo `json:"networkDeviceInfo,omitempty" protobuf:"bytes,7,rep,name=networkDeviceInfo"`

Reviewer 4: ```suggestion
    NetworkInfo NetworkInfo `json:"networkInfo,omitempty" protobuf:"bytes,7,rep,name=networkInfo"`
```
What people think?

Author: I think this makes sense, do you think the `DeviceInfo` field should be renamed to `Info`?

Reviewer: I think "Opaque" is the word we use elsewhere for RawExtension from fields. That makes it clear only the driver should have expectations about what is in there.

Author: In the device configuration, `Opaque` is a combination of the driver + parameters:
https://github.com/kubernetes/api/blob/v0.31.1/resource/v1alpha3/types.go#L579

Reviewer: Oh, right, it's the name of the one-of. Can you check other RawExtension fields in the API? Is there any de facto convention?

Author: I have only seen the usage of RawExtension called `Data` in [apps/v1](https://github.com/kubernetes/api/blob/v0.31.1/apps/v1/types.go#L953)

Otherwise, there are some other usages of it called `Object`, `Parameters`, `Items`, `Config`, `Extension`, `Args`, `Data`

Reviewer: Ok. `Data` here is better than `Info`, IMO.

Author: Nobody seemed against it, so I renamed the field and struct `NetworkDeviceInfo` to `NetworkDeviceData` and the field `Info` to `Data` as suggested.

Reviewer: sgtm

}

// NetworkDeviceInfo provides network-related details for the allocated device.
// This information may be filled by drivers or other components to configure
// or identify the device within a network context.

> type NetworkDeviceInfo struct {

Reviewer 5: This comment seems to mean "does 'network device' mean a whole 'network'?", at least to me. At least to me, I suppose it does not.

'A network device' is a device that connects to 'a network', so a network device could be just a part of ' a network'.

Other cloud infrastructures tend to use 'network interface' for such network device information, such as
- https://cloud.google.com/vpc/docs/create-use-multiple-interfaces#creating_virtual_machine_instances_with_multiple_network_interfaces
- https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-network-interfaces.html

hence, user may get confused if Kubernetes uses the term, 'network', for 'network interface' information because they may expect 'network interface' or 'network device' for such information.

If we call 'NetworkInfo', then it should contain 'network address', not 'network interface address' because it is 'NetworkInfo'.

> type NetworkDeviceInfo struct {

Reviewer: Well..."NetworkInfo" could mean "the network info associated with this device"...since we are already narrowed down the context to a device.

    // Interface specifies the name of the network interface associated with
    // the allocated device. This might be the name of a physical or virtual
    // network interface.
    //
    // +optional

>     Interface string `json:"interface,omitempty" protobuf:"bytes,1,rep,name=interface"`

Reviewer 4: this has to be `InterfaceName` to me to be clear that is the `name` , that is also the kernel identifier for the network interfaces https://man7.org/linux/man-pages/man7/netdevice.7.html

Author: Right, I renamed it, thank you.

    // NetworkAddresses lists the network addresses assigned to the device's network interface.
    // This can include both IPv4 and IPv6 addresses.
    //
    // +optional

>     NetworkAddresses []NetworkAddress `json:"networkAddresses,omitempty" protobuf:"bytes,2,rep,name=networkAddresses"`

Reviewer 4: I think the `Network` prefix here is redundant, if we are going to call this `NetworkInfo.NetworkAddresses` reads much better `NetworkInfo.Addresses` 

I think is a good idea to avoid to mention IP so we don't have to worry in the future about people wanting to use different protocols

Author: I renamed it `Addresses` as suggested.

    // Mac represents the MAC address of the device's network interface.
    //
    // +optional

>     Mac string `json:"mac,omitempty" protobuf:"bytes,3,rep,name=mac"`

Reviewer 4: we we have an use case for this? if we are going to make this generic , then this should be HWAddress

Author: I don't have specific use-case for it except troubleshooting. I renamed it to `HWAddress`.

}

// NetworkAddress provides a network address related details such as IP and Mask.

> type NetworkAddress struct {

Reviewer 4: why the indirection?

Author: I don't have specific use case for it, just in the future if we want to have the broadcast, unicast address and the scope, then we could still do it via this struct. Otherwise we could just have a `type CIDR string` instead.

Reviewer: we should follow the same conventions we follow elsewhere, if at all possible

Reviewer 4: we use `string` and `[]strings` for ips and cidrs, we should have `Addresses []strings`
If there is need to add new fields they can be added to the original struct,

## Design Details

### API

The `ResourceClaimStatus` struct in `pkg/apis/resource/types.go` will be
extended to include the slice of `DeviceStatuses`.

`ResourceClaim` validation of the status in
`pkg/apis/resource/validation/validation.go` will be covered to allow a device
to be reported only once in the slice, a device is being identified  by
`<driver name>/<pool name>/<device name>`.

### Test Plan

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

##### Unit tests

- `ResourceClaim` validation:
    - A device can only be reported once in the `ResourceClaim`.
    - The reported device is allocated in the `ResourceClaim`.
    - Properties set in `AllocatedDeviceStatus` are in the correct format.

##### Integration tests

- Usage of the `DeviceStatuses` field in the `ResourceClaimStatus`:
    * With the feature gate enabled, the field exists in the `ResourceClaim`.
    * With the feature gate disabled, the field does not exist in the
      `ResourceClaim`.

##### e2e tests

> TBD

Reviewer 4: we need to work on this, we need to create a network DRA driver and use dummy interfaces per example, that is how I'm doing these days.

Reviewer 4: at minimum we implement an e2e test that populate these new fields using the existing test driver https://github.com/kubernetes/kubernetes/tree/master/test/e2e/dra

at best we have an e2e test with a dra network driver that add multiple interfaces

Author: I added those suggestion in the latest commit, thank you.

### Graduation Criteria

#### Alpha

> - Feature implemented behind feature gates (`ResourceClaimDeviceStatus`).

Reviewer 4: do we want to start a new track for adding a new field? is not possible to reuse existing track @johnbelamaric @pohly @klueska

Reviewer 2: I don't think it fits into some other track. It's a new feature with new fields specifically for it.

Reviewer 2: Do we want to start DRA-related feature gates with "DRA"? I did for `DRAControlPlaneController` (= classic DRA).

Reviewer 4: ack

Reviewer: Yeah, we can't include it in another feature gate and still move forward independently.

Author: I renamed it as `DRAControlPlaneController` as suggested.

Reviewer 2: I wasn't suggesting to use exactly that - it's already taken! :sweat_smile: 

Perhaps use `DRAClaimDeviceStatus`?

Author: Ho yes, sorry, I wanted to say `DRAResourceClaimDeviceStatus`, not `DRAControlPlaneController`

No

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

Depending on the content of the `Info` set by the DRA drivers, the disk
usage could increase.

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

No

### Troubleshooting

###### How does this feature react if the API server and/or etcd is unavailable?

N/A

###### What are other known failure modes?

N/A

###### What steps should be taken if SLOs are not being met to determine the problem?

N/A

## Implementation History

- Initial proposal: 2024-08-30

## Drawbacks

If the Network device (network interface) characteristics (e.g. IP) and status
is reported as part of the `Resource.Claim.Status`, it should be ensured the
`ResourceClaim` is not used by several `Pod` at a time. Additionally, if a
controller needs to gather IPs for a specific network to which Pods are
attached via networking devices, it will need to query each `Pod` and then
access the corresponding `ResourceClaim` for every Pod.

## Alternatives

> ### Annotations

Reviewer: CustomResources are another alternative

extended to include the slice of `Devices`.

`ResourceClaim` validation of the status in
`pkg/apis/resource/validation/validation.go` will be covered to allow a device
to be reported only once in the slice, a device is being identified  by
`<driver name>/<pool name>/<device name>`.

### Write Permission

To prevent unauthorized or accidental modifications by entities that do not 
have access to a particular resource, a `ValidatingAdmissionPolicy` will be 
created to validate the entities attempting to update the devices in the 
`ResourceClaim.Status`.

The `ValidatingAdmissionPolicy` will restrict `ResourceClaim.Status.Devices` 
to be set only during updates, as the object will have first to be created and 
allocated, then configured inside the pods. It will also restrict the 
`ResourceClaim.Status.Devices` to be set only for when the `ResourceClaim` is 
allocated to a node. Additionally, the allocated node where the `ResourceClaim` 
is assigned will be used to check if the user/entity updating the 
`ResourceClaim.Status.Devices` is running on the same node.

Here is a `ResourceClaim` allocated on a node. This would only work for now if 

> exactly one node is set:

Reviewer: What will happen if the node selector is not of the "one node" form?

Author: This was in the conversation on slack because it was not possible to check if the node matches the node selector. But I don't think there was any conclusion for this. @aojea @pohly

Reviewer: We need to answer it here. The answer is probably "we fail and reject the update". But it could be "that means node-based authorization isn't needed for this claim" and so we allow it. We need to take a stand for the first alpha, at least.

Reviewer 2: It has to fail close. Otherwise we might accidentally grant all node-agents write permission for something that they shouldn't have access to.

Author: I added a paragraph about this in the last commit

Reviewer: awesome, thanks

Reviewer 6: I'm not sure I follow - this seems very rigid?  Any driver that doesn't follow this exact standard can't be secured?  Or do we end up embedding thdriver name into the VAP, and have one VAP per driver?

Reviewer: yes, I don't love this either, but I figured it's OK for alpha. For beta, I would either want to figure out how to have native, pre-loaded VAP + VAP that can test a label selector, or do an admission plugin.

Reviewer 4: I read the VAP thing as optional/recommended, I imagine different drivers will provide different policies depending on the use cases, but this seems to fit in the model existing network device drivers work

    // identify which device was allocated (`<driver name>/<pool name>/<device name>`).
    //
    // Must not be longer than 253 characters and may contain one or more
    // DNS sub-domains separated by slashes.
    //
    // +required
    Pool string `json:"pool" protobuf:"bytes,2,rep,name=pool"`

    // Device references one device instance via its name in the driver's
    // resource pool. It must be a DNS label.
    //
    // +required
    Device string `json:"device" protobuf:"bytes,3,rep,name=device"`

    // Conditions contains the latest observation of the device's state.
    // If the device has been configured according to the class and claim
    // config references, the `Ready` condition should be True.
    //
    // +optional
    // +listType=atomic
    Conditions []metav1.Condition `json:"conditions" protobuf:"bytes,4,opt,name=conditions"`

    // Data contains arbitrary driver-specific data.
    //
    // +optional
    // +listType=atomic
    Data []runtime.RawExtension `json:"data,omitempty" protobuf:"bytes,5,opt,name=data"`

    // NetworkData contains network-related information specific to the device.
    //
    // +optional

>     // +oneOf=DeviceDataType

Reviewer: this is not a oneOf any more in this version

Author: I just removed it

    // vendor of the driver.
    //
    // +required
    Driver string `json:"driver" protobuf:"bytes,1,rep,name=driver"`

    // This name together with the driver name and the device name field
    // identify which device was allocated (`<driver name>/<pool name>/<device name>`).
    //
    // Must not be longer than 253 characters and may contain one or more
    // DNS sub-domains separated by slashes.
    //
    // +required
    Pool string `json:"pool" protobuf:"bytes,2,rep,name=pool"`

    // Device references one device instance via its name in the driver's
    // resource pool. It must be a DNS label.
    //
    // +required
    Device string `json:"device" protobuf:"bytes,3,rep,name=device"`

    // Conditions contains the latest observation of the device's state.
    // If the device has been configured according to the class and claim
    // config references, the `Ready` condition should be True.
    //
    // +optional
    // +listType=atomic
    Conditions []metav1.Condition `json:"conditions" protobuf:"bytes,4,opt,name=conditions"`

    // Data contains arbitrary driver-specific data.
    //
    // +optional
    // +listType=atomic

>     Data []runtime.RawExtension `json:"data,omitempty" protobuf:"bytes,5,opt,name=data"`

Reviewer 4: do we need this to be an array? 
```
grep -r \\[\\]runtime.RawExtension staging/src/k8s.io/api
staging/src/k8s.io/api/core/v1/zz_generated.deepcopy.go:                *out = make([]runtime.RawExtension, len(*in))
```
is a rawextension, so it can use  complex structures inside,  let's make this is a singular or is there any special motivation for making it an array?

Author: The main motivation was to report different types of data, like the [cni result](https://github.com/containernetworking/cni/blob/v1.2.3/pkg/types/100/types.go#L90), and the [multus network status](https://github.com/k8snetworkplumbingwg/network-attachment-definition-client/blob/v1.7.1/pkg/apis/k8s.cni.cncf.io/v1/types.go#L103).

cc @mskrocki @dougbtv

Reviewer 4: I don't think we need to leak those details into the API, we are standardising the common fields, if some implementation want to use custom data  it has the RawExtension and model it the way they want, IIUIC it is perfectly fine to add those both types in the same Data field, and also we avoid having two actors on the same resource.
I'm not in favor of an array here and, in the worst scenario, if we find out that is needed, pluralising the field is not a breaking change,

Author: I changed it to a single RawExtension as suggested
```golang
type AllocatedDeviceStatus struct {
    ...
    // Data contains arbitrary driver-specific data.
    //
    // +optional
    Data runtime.RawExtension `json:"data,omitempty" protobuf:"bytes,5,opt,name=data"`
    ...
}
```

>     Data []runtime.RawExtension `json:"data,omitempty" protobuf:"bytes,5,opt,name=data"`

Reviewer 4: ```suggestion
    Data runtime.RawExtension `json:"data,omitempty" protobuf:"bytes,5,opt,name=data"`
```

    // NetworkData contains network-related information specific to the device.
    //
    // +optional
    NetworkData NetworkDeviceData `json:"networkData,omitempty" protobuf:"bytes,6,opt,name=networkData"`
}

// NetworkDeviceData provides network-related details for the allocated device.
// This information may be filled by drivers or other components to configure
// or identify the device within a network context.
type NetworkDeviceData struct {
    // InterfaceName specifies the name of the network interface associated with
    // the allocated device. This might be the name of a physical or virtual
    // network interface.
    //
    // +optional
    InterfaceName string `json:"interfaceName,omitempty" protobuf:"bytes,1,opt,name=interfaceName"`

>     // Addresses lists the network addresses assigned to the device's network interface.
>     // This can include both IPv4 and IPv6 addresses.
>     //
>     // +optional
>     // +listType=atomic
>     Addresses []NetworkAddress `json:"addresses,omitempty" protobuf:"bytes,2,opt,name=addresses"`

Reviewer 4: ```suggestion
    // Addresses lists the network addresses assigned to the device's network interface.
    // This can include both IPv4 and IPv6 addresses.
    // Addresses contains the network address in CIDR notation, which includes
    // both the address and the associated subnet mask.
    // e.g.: "192.0.2.5/24" for IPv4 and "2001:db8::9/64" for IPv6.
    //
    // +optional
    // +listType=atomic
    Addresses []string `json:"addresses,omitempty" protobuf:"bytes,2,opt,name=addresses"`
```

### Risks and Mitigations

As stated, 3rd party DRA drivers will set and update the `Devices` for
the device they manage. An access control must be set in place to restrict the
write access to the appropriate driver (A device status can only be updated by
the entities that have a direct control over the device(s) being reported).

Adding `Data` as an arbitrary data slice may introduce extra processing
and storage overhead which might impact performance in a cluster with many
devices and frequent status updates. In large-scale clusters where many devices
are allocated, this impact must be considered.

## Design Details

### API

The `ResourceClaimStatus` struct in `pkg/apis/resource/types.go` will be
extended to include the slice of `Devices`.

`ResourceClaim` validation of the status in
`pkg/apis/resource/validation/validation.go` will be covered to allow a device
to be reported only once in the slice, a device is being identified  by
`<driver name>/<pool name>/<device name>`.

### Write Permission

To prevent unauthorized or accidental modifications by entities that do not 
have access to a particular resource, a `ValidatingAdmissionPolicy` will be 

> created to validate the entities attempting to update the devices in the 

Reviewer 4: just curios, is this created at cluster startup? using the `plugin/pkg/auth/authorizer`?

Reviewer 2: No, this is something that a cluster admin has to add once the cluster is up and running.

I have a topic pending for SIG Arch about automatically managing VAPs that are part of core Kubernetes. But as for CRDs, there's probably not going to be a good solution.

      restricts the write access of the `ResourceClaim.Status.Devices`.

##### e2e tests

TBD

### Graduation Criteria

#### Alpha

- Feature implemented behind feature gates (`DRAResourceClaimDeviceStatus`).
  Feature Gates are disabled by default.
- Documentation provided.
- Initial unit, integration and e2e tests completed and enabled.
- Authorization implemented to allow only the user on the same node as the 
  allocated `ResourceClaim` to write the status of the devices.

#### Beta

- Feature Gates are enabled by default.
- No major outstanding bugs.
- Feedback collected from the community (developers and users) with adjustments
  provided, implemented and tested.

#### GA

- 2 examples of real-world usage.
- Allowing time for feedback from developers and users.

### Upgrade / Downgrade Strategy

This feature only exposes a new field in the `ResourceClaim.Status`, the field
will either be present or not.

DRA implementation requires DRA interfaces change. DRA is in alpha and in
active development. The feature will follow the DRA upgrade/downgrade strategy.

### Version Skew Strategy

This feature affects only the kube-apiserver, so there is no issue with version

> skew with other Kubernetes components.

Reviewer: There could be skew with the DRA drivers themselves, but I guess that's their problem?

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: DRAResourceClaimDeviceStatus
  - Components depending on the feature gate: kube-apiserver
- [ ] Other
  - Describe the mechanism:
  - Will enabling / disabling the feature require downtime of the control
    plane?
  - Will enabling / disabling the feature require downtime or reprovisioning of
    a node?

###### Does enabling the feature change any default behavior?

No

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes, with no side effects except for missing the new field in
`ResourceClaim.Status`. Re-enabling this feature will not guarantee to keep

> the values written before the feature has been disabled.

Reviewer: what happens to the DRA plugins that are trying to write the field?

Can you provide guidance for plugin authors on how to handle this?

Author: I am not sure what else to describe except the field would be missing.
I added the DRA-Driver will have to re-write the status in case the feature is re-enabled.

###### What happens if we reenable the feature if it was previously rolled back?

> 

Reviewer: Just say that the field is available again for read and write.

###### Are there any tests for feature enablement/disablement?

Enablement/disablement of this feature is tested as part of the integration
tests.

### Rollout, Upgrade and Rollback Planning

> No

Reviewer: ?

###### How can a rollout or rollback fail? Can it impact already running workloads?

> No

Reviewer: You probably want to say something like:

Enabling the feature will enable the field to be written and therefore invoke validation of the field. Since the feature does nothing more than allow storage of this data, the only way it can really fail is to crash in validation. It will not affect running workloads.

###### What specific metrics should inform a rollback?

N/A

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

N/A

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

No

### Monitoring Requirements

###### How can an operator determine if the feature is in use by workloads?

> Check the `ResourceClaim.Status.Devices`.

Reviewer: Typically we want a metric for this. A count of ResourceClaims with this field set? Can you get that from kube state metrics?

Author: resourceclaim_status_devices_update_attempts_total and resourceclaim_status_devices_update_failures_total have been added. I am not sure about kube state metrics, I could take a look at it, but this will be directly implemented in kube-apiserver right?

###### How can someone using this feature know that it is working for their instance?

- [ ] Events
  - Event Reason: 
- [x] API .status
  - Condition name: 
  - Other field: `ResourceClaim.Status.Devices`
- [ ] Other (treat as last resort)
  - Details:

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

N/A

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

- [ ] Metrics
  - Metric name:
  - [Optional] Aggregation method:
  - Components exposing the metric:
- [ ] Other (treat as last resort)
  - Details:

###### Are there any missing metrics that would be useful to have to improve observability of this feature?

<!--
Describe the metrics themselves and the reasons why they weren't added (e.g., cost,
implementation difficulties, etc.).
-->

### Dependencies

[KEP-4381 - DRA Structured Parameters](https://github.com/kubernetes/enhancements/issues/4381)

###### Does this feature depend on any specific services running in the cluster?

> No

Reviewer: No, but the field won't be populated unless a DRA driver utilizes it.

    * With the feature gate enabled, the field exists in the `ResourceClaim`.
    * With the feature gate disabled, the field does not exist in the
      `ResourceClaim`.
    * With the feature gate enabled, the `ValidatingAdmissionPolicy` exists and
      restricts the write access of the `ResourceClaim.Status.Devices`.

##### e2e tests

The [DRA test driver](https://github.com/kubernetes/kubernetes/tree/master/test/e2e/dra/test-driver)
will be extended to support the new  `ResourceClaim.Status.Devices` field and to 
populate it with data related to the device configured in the pod.

A new network DRA Driver will be implemented (or extended from the existing DRA 
test driver) to support networking type of devices and report their 
network status.

### Graduation Criteria

#### Alpha

- Feature implemented behind feature gates (`DRAResourceClaimDeviceStatus`).
  Feature Gates are disabled by default.
- Documentation provided.
- Initial unit, integration and e2e tests completed and enabled.
- Authorization implemented to allow only the user on the same node as the 
  allocated `ResourceClaim` to write the status of the devices.

> #### Beta

Reviewer 4: Talked with @danwinship offline, @LionelJouin @dougbtv what do you think if we add to beta criteria having a multus DRA network driver ?

Reviewer 7: I'd definitely be interested in it! And there's a sort of basis that exists pre-1.31 in: https://github.com/k8snetworkplumbingwg/multus-cni/pull/1078

But yes, it'd need extensions to this functionality as well as a Multus DRA driver itself, as the challenge with that implementation is that it uses an example reference driver, and directly being a driver would be a good way to see how well the spec fits with an implementation.

Author: I added 1 real-world usage as beta requirement.

Enabling the feature will enable the field to be written and therefore invoke 
validation of the field. Since the feature does nothing more than allow storage 
of this data, the only way it can really fail is to crash in validation. It 
will not affect running workloads.

###### What specific metrics should inform a rollback?

N/A

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

N/A

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

No

### Monitoring Requirements

###### How can an operator determine if the feature is in use by workloads?

Check the `ResourceClaim.Status.Devices`.

###### How can someone using this feature know that it is working for their instance?

- [ ] Events
  - Event Reason: 
- [x] API .status
  - Condition name: 
  - Other field: `ResourceClaim.Status.Devices`
- [ ] Other (treat as last resort)
  - Details:

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

N/A

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

- [x] Metrics

>   - Metric name: `resourceclaim_status_devices_update_attempts_total` ; `resourceclaim_status_devices_update_failures_total`

Reviewer: nit (won't block merge): can you make this a yaml-style list? Also note the first one above in the "how can an operator determine if the feature is in use"

Author: Done

characteristics of what has been allocated is particularly useful and crucial.
For example, reporting the interface Name, MAC address and IP addresses of
network interfaces in the status of a ResourceClaim can significantly help in
configuring and managing network services, as well as in debugging network
related issues. By including such device specific information, this proposal
addresses existing gaps in visibility and facilitates better integration and
management of resources.

### Goals

* Allow arbitrary, driver-specific information to be reported from the DRA
  drivers for each allocated device in a ResourceClaim.
* Establish a standardization for device status (e.g. Network Devices status).
* Enable 3rd party implementation of new functionalities based on the Device
  Status (e.g. Secondary Network Services if the IPs of a network device are
reported).

### Non-Goals

* Implement new functionalities based on the Device Status (e.g. Kubernetes
  EndpointSlice Controller supporting IPs reported by networking DRA drivers).
* Modifying the Resource Claim workflow to include the device status report.

## Proposal

> ### API - ResourceClaim.Status

Reviewer 8: The difference between "Proposal" and "Design Details" has always been confusing to me, but in this case it's especially confusing that the details of the proposal appear before the user stories, so I think this should mostly move to "Design Details".

`NetworkData` can be used to report the IPs, the MAC address and the
interface name.

`Data` being an arbitrary data field allows the DRA Driver to store
device specific data in different formats. For example, a Network Device being
configured by via a CNI plugin could get its `Data` field filled with the CNI
result for troubleshooting purpose and with a Network result in a modern
standard format (closer to Pod.Status.PodIPs for example) used by 3rd party
controllers.

For each device, if required, the DRA Driver processing the device allocation
can report the status of it in the `Status.Devices` of the ResourceClaim
by using the Kubernetes API.

```golang
// ResourceClaimStatus tracks whether the resource has been allocated and what
// the result of that was.
type ResourceClaimStatus struct {
    ...

    // Devices contains the status of each device allocated for this
    // claim, as reported by the driver. This can include driver-specific
    // information. Entries are owned by their respective drivers.
    //
    // +optional
    // +listType=map
    // +listMapKey=driver
    // +listMapKey=device
    // +listMapKey=pool

>     // +featureGate=DRAResourceClaimDeviceStatus

Reviewer: yes, different drivers will write different device statuses

>     // +featureGate=DRAResourceClaimDeviceStatus

Reviewer: To be clear: A single driver may write multiple entries in here, and multiple drivers may write to this map, but any given device will have no more than one entry, and any given device is owned by one driver (so any given entry is owned by one driver).

    Devices []AllocatedDeviceStatus `json:"devices,omitempty" protobuf:"bytes,4,opt,name=devices"`
}

// AllocatedDeviceStatus contains the status of an allocated device, if the
// driver chooses to report it. This may include driver-specific information.

> type AllocatedDeviceStatus struct {

Reviewer: Yes, a request can have more than one device. We need the device triple here; we can map that back to the request if we really need to via the allocation field, I believe.

> type AllocatedDeviceStatus struct {

Reviewer 6: OK, then just remove the part saying it will mention request :)

> type AllocatedDeviceStatus struct {

Author: I did remove the request part in the last commit, thank you.

    // Driver specifies the name of the DRA driver whose kubelet
    // plugin should be invoked to process the allocation once the claim is
    // needed on a node.
    //
    // Must be a DNS subdomain and should end with a DNS domain owned by the
    // vendor of the driver.
    //
    // +required
    Driver string `json:"driver" protobuf:"bytes,1,rep,name=driver"`

    // This name together with the driver name and the device name field
    // identify which device was allocated (`<driver name>/<pool name>/<device name>`).
    //
    // Must not be longer than 253 characters and may contain one or more
    // DNS sub-domains separated by slashes.
    //
    // +required

>     Pool string `json:"pool" protobuf:"bytes,2,rep,name=pool"`

Reviewer 8: doc comment should start with "Pool", right?

Also, just to confirm: some pieces of code will be using the string `Driver + "/" + Pool + "/" + Device`, but the pool name itself is allowed to contain internal slashes?

Author: Right, I copied it from the resource API in the [DeviceRequestAllocationResult struct](https://github.com/kubernetes/api/blob/v0.31.1/resource/v1alpha3/types.go#L759)

    // Device references one device instance via its name in the driver's
    // resource pool. It must be a DNS label.
    //
    // +required
    Device string `json:"device" protobuf:"bytes,3,rep,name=device"`

>     // Conditions contains the latest observation of the device's state.

Reviewer: Yes, other than `Ready`, it's all up to the driver what lands in here. This will be the only place to put that information in the API server (we will not have it in pod status, for example).

>     // Conditions contains the latest observation of the device's state.

Reviewer: well, conditions are just boolean. standard guidance https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties for conditions would apply.

>     // Conditions contains the latest observation of the device's state.

Author: The guidance on what should go in `Conditions` and `Data` are in the "Proposal" section:
```yaml
The state and characteristics of the
device are reported in the `Conditions`, representing the operational state of
the device and in the `Data`, an arbitrary data field representing device
specific characteristics. 
....

`Data` being an arbitrary data field allows the DRA Driver to store
device specific data in different formats. For example, a Network Device being
configured by via a CNI plugin could get its `Data` field filled with the CNI
result for troubleshooting purpose and with a Network result in a modern
standard format (closer to Pod.Status.PodIPs for example) used by 3rd party
controllers.
```
or did you mean more detailed ones?

>     // Conditions contains the latest observation of the device's state.

Reviewer 6: Conditions are not an awesome API for most things, but they are the only way to express things without knowing the "extension" schema.

I suspect we will have to feel this distinction out as we go, and crystallize it more later.

    // If the device has been configured according to the class and claim
    // config references, the `Ready` condition should be True.
    //
    // +optional
    // +listType=atomic
    Conditions []metav1.Condition `json:"conditions" protobuf:"bytes,4,opt,name=conditions"`

    // Data contains arbitrary driver-specific data.
    //
    // +optional
    Data runtime.RawExtension `json:"data,omitempty" protobuf:"bytes,5,opt,name=data"`

    // NetworkData contains network-related information specific to the device.
    //
    // +optional

>     NetworkData NetworkDeviceData `json:"networkData,omitempty" protobuf:"bytes,6,opt,name=networkData"`

Reviewer 7: Question about `NetworkData` vs. `Data`. So let's consider a situation where I've got users using NetworkAttachmentDefinition style specifications, and so they've got fields like:

```
    Name       string      `json:"name"`
    Interface  string      `json:"interface,omitempty"`
    IPs        []string    `json:"ips,omitempty"`
    Mac        string      `json:"mac,omitempty"`
    Mtu        int         `json:"mtu,omitempty"`
    Default    bool        `json:"default,omitempty"`
    DNS        DNS         `json:"dns,omitempty"`
    DeviceInfo *DeviceInfo `json:"device-info,omitempty"`
    Gateway    []string    `json:"gateway,omitempty"`
```

The so, I have this mapping...

```
    Interface  --> InterfaceName
    IPs  --> Addresses
    Mac --> HWAddress
```

Which is 3/x, so the other values would be remapped to... The data field? 

I'm mostly looking for clarity, and trying to figure out standardization (where it makes sense), and also -- extensibility. That is, that with the `*/network-status:` annotation with `net-attach-def`, we've definitely had to slowly expand that struct over time. So I wonder if we should focus on building some standards for the `Data` field if these other fields would go there? Or, could we have something as part of the `NetworkDeviceData` that is somehow extensible, and maybe think about forming a standard for it via Multinetworking WG?

Reviewer: If you want to store the rest, yes, the `Data` field. The main idea is that the Data field is 100% driver specific and so there are no guarantees (by the Kubernetes project) with respect to this data structure or API evolution.

The question I have is what data needs to be consumed by other ecosystem projects? That's what should be in a K8s built-in API.

> Or, could we have something as part of the NetworkDeviceData that is somehow extensible, and maybe think about forming a standard for it via Multinetworking WG?

Maybe `NetworkDeviceData` should be a K8s one-of?

Or, to be really flexible, instead of two fields, we have one field `Data []DeviceData`, sort of like we did originally but now an array instead of just a single value:

```go
type AllocatedDeviceStatus struct {
 ...
 Data []DeviceData
}

type DeviceData struct {
  // exactly one must be populated

  // +optional
  Opaque *runtime.RawExtension

  // +optional
  NetworkData *NetworkDeviceData

  // +optional
  SomeFutureThing *MultinetThing
}
```

The sort of hard part is now we can have an aribtrary number of each type, but maybe that's OK.

Reviewer 4: @dougbtv one question, what of those things are configuration option and what of those things need to return a value from the driver?

It look most of those things are configuration options, that already are reported in status, the thing we try to standardize are the things that are reported by the driver that are not set in configuration.

Reviewer 8: The earlier text implies that `Data` may include multiple representations of the same data in different formats:

>     `Data` being an arbitrary data field allows the DRA Driver to store
>     device specific data in different formats. For example, a Network Device being
>     configured by via a CNI plugin could get its `Data` field filled with the CNI
>     result for troubleshooting purpose and with a Network result in a modern
>     standard format (closer to Pod.Status.PodIPs for example) used by 3rd party
>     controllers.

Or maybe that's an artifact from an earlier version of the text and the second part is supposed to be the `NetworkData` field?

If there is any intention that `Data` could have multiple representations of the same data, then it seems like it ought to be explicitly an array or map. Otherwise how are you supposed to be able to reliably extract the representation you wanted?

So I think I like John's version above.

Reviewer: yes, this is definitely not for config - config goes in the `DeviceClaim.Config` list, which is a bunch of RawExtensions (or in the `DeviceClass.Config`).

Reviewer 4: Let me put an example on how it will look like (you can embed the json directly if you want in `opaque` but I always hear you complain about it 😄 

User configures network attachment definition in a 
```
apiVersion: resource.k8s.io/v1alpha3
kind:  ResourceClaim
metadata:
  name: multus-test-1
spec:
  devices:
    requests:
      selectors:
        - cel:
            expression: device.driver == "multus.dra.networking.driver"
    config:
    - opaque:
        driver: multus.dra.networking.driver
        parameters:
 	  cniVersion: 0.3.1
  	  type: macvlan
  	  master: ens2
  	  mode: bridge
  	  ipam:
    	  - type: whereabouts
    	     range: 192.168.1.0/24
```

Then it creates a Pod referencing it

```
---
apiVersion: v1
kind: Pod
metadata:
  name: pod1
  labels:
    app: pod
spec:
  containers:
  - name: ctr1
    image: registry.k8s.io/e2e-test-images/agnhost:2.39
  resourceClaims:
  - name: dummy1
    resourceClaimName: multus-test-1
```

The pod gets scheduled and now the ResourceClaim reports back in status

1. The config you have passed
2. The standardize NetworkData: IPs, mac, ... that can be used across the ecosystem to create endpoint slices controller per example
3. The opaque Data object with data that the specific vendor wants to report back from the claim allocation

Reviewer 6: I can see this a number of ways, depending on some answers to above questions.

Do we think that NetworkData is mutually-exclusive with data (i.e. one-of) ?  Or can they both be set at the same time?

"Data" is a hateful word.  I wonder if we can minimize it or find something better.

Option 1: a series of options, not exclusive.
```
    Data *runtime.RawExtension
    NetworkData *DeviceNetworkData
    FrobberData *DeviceFrobberData
```

Option 2: Mutually exclusive, but extgensible options.
```
type DeviceNetworkData struct {
    InterfaceName string
    Addresses []string
    Ext *runtime.RawExtension
}

type DeviceFrobberData struct {
    ID string
    Ext *runtime.RawExtension
}
```

Option 3: One of the above, but nested:
```
type ResourceClaimStatus struct {
  // ...
  Devices []AllocatedDeviceStatus
}

type AllocatedDeviceStatus struct {
  // ...
  Data AllocatedDeviceStatusData
}

type AllocatedDeviceStatusData struct {
    Opaque *runtime.RawExtension
    Network *DeviceNetworkData
    Frobber *DeviceFrobberData
}
```

Or some other arrangement?  We can fine-detail it in API review, but the question is mostly about clarity - which things are extensible, which are fixed, which are composed, and which are exclusive.

Reviewer: My suggestion is your 3) except with a slice:`Data []AllocatableDeviceStatusData`. This works for compound device drivers too, or devices that provide multiple functions.

Reviewer: > "Data" is a hateful word. I wonder if we can minimize it or find something better.

We poked around for other uses of `RawExtension` and that one came up a few times. Being a bit more thorough:

```console
[hi on] jbelamaric@jbelamaric:~/proj/gh/johnbelamaric/kubernetes/kubernetes$ grep RawExtension $(find . -name types.go) | grep -v // | cut -f 2 -d : | cut -f 1 -d ' ' | tr -d '\t ' | sort | uniq -c | sort -k 1nr
      4 Data
      4 Object
      3 Extension
      2 Config
      2 ConvertedObjects
      2 Items
      2 Objects
      2 OldObject
      2 Options
      2 Parameters
      1 Args
      1 EmptyObject
      1 rawExtension
      1 type
[hi on] jbelamaric@jbelamaric:~/proj/gh/johnbelamaric/kubernetes/kubernetes$ 
```

So...not sure `Object` or `Objects` or `Items` is better than `Data`. And it's not `Config` or `Parameters` or `Extensions` to me. `Info` seems at least as bad as `Data`. Other ideas?

Reviewer: It's not mutually exclusive multi-function devices. In particular, I want to build a driver that allocates multiple underlying devices as a single logical unit. Each of those has its own driver with its own output. 

And yes, DriverOutput sounds good to me

Reviewer 6: In context:
   `obj.status.devices[i].data.{opaque,network,whatever}`
   `obj.status.devices[i].driverOutput.{opaque,network,whatever}`
   `obj.status.devices[i].results.{opaque,network,whatever}`
   `obj.status.devices[i].info.{opaque,network,whatever}`

Staring at that, `data` seems not-so-bad.  I don't think `driverOutput` is better.  What about

   `obj.status.devices[i].driver.{opaque,network,whatever}`
or
  `obj.status.devices[i].status.{opaque,network,whatever}`

?

> My suggestion is your 3) except with a slice

So instead of: 

```
type AllocatedDeviceStatus struct {
    // ...

    // At least one of the following must be set. As this evolves
    // over time, back-rev clients might not know about new options,
    // and so might see none of the fields they know.  
    Data *runtime.RawExtension
    NetworkData *DeviceNetworkData
    FrobberData *DeviceFrobberData
}
```
or

```
type AllocatedDeviceStatus struct {
  // ...
  Data AllocatedDeviceStatusData
}

type AllocatedDeviceStatusData struct {
    // At least one of the following must be set. As this evolves
    // over time, back-rev clients might not know about new options,
    // and so might miss some data or see none of the fields they know.
    Opaque *runtime.RawExtension
    Network *DeviceNetworkData
    Frobber *DeviceFrobberData
}
```

You are saying:

```
type AllocatedDeviceStatus struct {
    // ...

    // Each item in this list is expected to have one sub-option set.
    // If a client finds an empty item in this list, they can infer that
    // there was some data that they do not understand.
    Data []AllocatedDeviceStatusData
}

type AllocatedDeviceStatusData struct {
    // Exactly one of the following must be set. As this evolves
    // over time, back-rev clients might not know about new options,
    // and so might see none of the fields they know.  
    Opaque *runtime.RawExtension
    Network *DeviceNetworkData
    Frobber *DeviceFrobberData
}
```

The main difference is that clients can now be aware that they are missing information.

a) We do this in other DRA places, but in this context it seems less critical (status vs spec) -- if the client needed that info, it would know about it.
b) We don't generally handle unknown fields well in clients, and it seems like a systemic solution would be better.

I am not against a list here - it's consistent with other DRA places.  Let's pick a model and name for the KEP and then revisit this detail in code review.

Reviewer 4: I see, I misinterpreted it sorry @dougbtv 

 @johnbelamaric is the list not going to be confusing to identify which information goes where?

Reviewer: > @johnbelamaric is the list not going to be confusing to identify which information goes where?

Yes, each list entry needs an identifier if we do this. Maybe make it a list-map?

> `obj.status.devices[i].driver.{opaque,network,whatever}`

`driver` already exists in `device[i]` as the name of the driver.

> `obj.status.devices[i].status.{opaque,network,whatever}`

Meh. Not better than `data` to me.

With data as a slice as I am suggesting:

`obj.status.devices[i].data[j].{opaque,network,whatever}`

Reviewer 6: > `obj.status.devices[i].data[j].{opaque,network,whatever}` 

So a claim has multiple devices, each device has multiple datas (datae?), each of which has exactly one field set, right?

Reviewer: yes, that's right. fun stuff

Reviewer: `data` is a list-map with some sort of driver-defined identifier for the key

>     NetworkData NetworkDeviceData `json:"networkData,omitempty" protobuf:"bytes,6,opt,name=networkData"`

Reviewer 6: pointers for optional fields

}

// NetworkDeviceData provides network-related details for the allocated device.
// This information may be filled by drivers or other components to configure

> // or identify the device within a network context.

Reviewer 8: Should probably include the interface index? Name, ips, and hwaddr are all mutable,.

(In which case _maybe_ you don't need to include the interface name?)

Reviewer 6: If you have enough privilege to rename an interface, none of the rest really matters, right?  That doesn't stop us from reporting podIPs

Reviewer 8: I wasn't worrying just about privilege/attacks. Maybe there are use cases where the interface eventually gets renamed for some reason?

It just seems like, if it's possible to refer to the device via an immutable identifier rather than a mutable one, then we should?

Reviewer 6: Agree with that principle

Reviewer 4: The reference of the interface in the kubernetes cluster is: Driver + "/" + Pool + "/" + Device
The index is the reference of the interface in a Linux operating system, Devices are supposed to abstract any network interface, most commonly the linux ones, but not only those.
The drivers owns the interface, if it is renamed, then I expect the driver to update the status, but I don't expect the consumer of the api to use the index of Linux interface, so interface name sounds much better to me

type NetworkDeviceData struct {
    // InterfaceName specifies the name of the network interface associated with
    // the allocated device. This might be the name of a physical or virtual
    // network interface.
    //
    // +optional

>     InterfaceName string `json:"interfaceName,omitempty" protobuf:"bytes,1,opt,name=interfaceName"`

Reviewer 8: why is this optional?

Reviewer 4: I think there cases that you want to report the addresses but not the interface name, I can think a few cases with network attached devices

Reviewer 7: DPDK / userspace stuff might not have an interface representation

    // Addresses lists the network addresses assigned to the device's network interface.
    // This can include both IPv4 and IPv6 addresses.
    // The addresses are in the CIDR notation, which includes both the address and the
    // associated subnet mask.

>     // e.g.: "192.0.2.0/24" for IPv4 and "2001:db8::/64" for IPv6.

Reviewer 8: So first, those examples are wrong; those are descriptions of subnets, not descriptions of IPs on those subnets. (You need, eg, "192.0.2.5/24").

But also, (1) this is contrary to everything else in kubernetes, where we always use plain IP addresses when referring to IP addresses, and only use CIDR notation for describing subnets and masks; (2) the CIDR notation for interface addresses is only really useful for IPv4 anyway, and doesn't really make sense for IPv6; (3) it is not clear what the user stories would be that would motivate including the link-local-subnet-mask but not any other routing information.

So I think these should be plain IPs.

Reviewer 6: Also we say "IPs" almost everywhere, and where we use "Addresses" it's because it could include non-IP values.  Which is this?

Reviewer 9: "192.0.2.0" and "2001:db8::" _are_ valid IP addresses.  Just try:
```
vm-003 ~ # ip -6 addr add 2001:db8::/64 dev eth0
vm-003 ~ # ip -6 addr show dev eth0
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    inet6 2001:db8::/64 scope global 
       valid_lft forever preferred_lft forever
    inet6 fd00::c0a8:3/120 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::200:ff:fe01:3/64 scope link proto kernel_ll 
       valid_lft forever preferred_lft forever
```
It is a common misunderstanding that host 0 (zero) is reserved. It isn't. But that misunderstanding may be buit-in in older systems/equipment.

Reviewer 9: But because of that misunderstanding 192.0.2.5/24 may be clearer

Reviewer 9: CIDR notation *does* make sense for IPv6. /64 is _not_ required, except for some generated addresses such as the link-local address and SLAAC. IMO it make perfect sense to have *one* /64 address for an entire K8s cluster, and divide it into POD, ClusterIP, and external addresses with CIDR

Reviewer 9: That said, I kind of agree that the mask should be omitted

Reviewer 8: Who is "you" here? The DRA driver? The end user? What is the _exact_ thing that someone can do when the prefix length is exposed here (without needing any other information that is also not exposed here) that they could not do with the prefix length not exposed.

Knowing the subnet size does not tell you how many IPs are _available_ on the subnet so I don't see how it helps with that.

Author: The end user can get this information and use it for example to get some hints on who (pod) can talk to whom (pod/external device). The CIDR could also tell you if what you got is what you requested (I requested an IP in the 192.0.2.0/24 subnet, is my address 192.0.2.x/24, or is there something wrong somewhere and my address is 192.0.2.5/8)

This `NetworkDeviceData` struct is likely to evolve in the future when people will start to solve more complex use-cases (as the `NetworkStatus` evolved in the [network-attachment-definition API](https://github.com/k8snetworkplumbingwg/network-attachment-definition-client/blob/v1.7.4/pkg/apis/k8s.cni.cncf.io/v1/types.go#L103) (Multus)). So I would prefer to have the addresses in the CIDR format now than just IP addresses and potentially a new field in the future for the CIDR.

    //
    // +optional
    // +listType=atomic

>     Addresses []string `json:"addresses,omitempty" protobuf:"bytes,2,opt,name=addresses"`

Reviewer 8: Are these allowed/expected to change while the ResourceClaim is claimed? That is, do these represent the _initially-assigned_ IPs, or the _current_ IPs?

    // HWAddress represents the hardware address (e.g. MAC Address) of the device's network interface.
    //
    // +optional

>     HWAddress string `json:"hwAddress,omitempty" protobuf:"bytes,3,opt,name=hwAddress"`

Reviewer 8: So the use case for including IP addresses in the result is that people may want to connect to the pod via this interface. That's something someone might want to do even if they aren't directly poking around with the allocated device.

What is the use case for including HW addresses? What sort of user would have a reason to need to know the hwaddr, but would _not_ be doing something where they had to look directly at the allocated device in question (which would allow them to find the hwaddr themselves)?

Reviewer 9: I am not directly involved any more, but I think non-standard (even seemingly weird) use-cases that don't use ip-addresses should be supported. An example is non-IP traffic, another is to allow "something" to send ip-packets to the interface even when it has no ip-address, or has an address shared by many.

The entity that consumes this status will probably not be an application that wants to connect to the pod, but something very specialized (router, load-balancer, ...?), and we can't predict what they will be.

}
```

> ### User Stories (Optional)

Reviewer 8: (you can remove the "(Optional)" from the section names where you're using that section)

#### Story 1 - Network Device Status for Network Services

As a Cloud Native Network Function (CNF) vendor, my network services must be
integrated with the network devices configured in Pods. The configuration
properties of these network devices are therefore essential to configure the
network services. For example, the network services must be able to route
traffic to pods over networks attached via the network devices, the IP
addresses of the network device(s) must then be reflected in the
`ResourceClaim.Status` allowing the network service controller(s) to access

> them.

Reviewer 8: So this motivates having the IPs be available, but not the hardware addrs (or anything else). Is that correct?

#### Story 2 - Network Device Status for Troubleshooting

As a Network Administrator, troubleshooting networking issues can be complex
and time consuming especially when the device characteristics and operational
status are not readily accessible. The `Devices` field in the
`ResourceClaim.Status` provides access to comprehensive details regarding
network interfaces helping to quickly and efficiently identify the issues such
as error messages on failed network interface configuration, incorrect IP
assignments or misconfigured network interfaces.

### Notes/Constraints/Caveats (Optional)

The content of `Data` is driver specific and not standardized as part of
DRA, the interpretation of this field may then vary between controllers and
users reading it. 

The accuracy of the information depends on the implementation of the DRA
Drivers, the reported status of the device may not always reflect the real time
changes of the state of the device.

### Risks and Mitigations

> As stated, 3rd party DRA drivers will set and update the `Devices` for

Reviewer 6: So the driver itself writes this, and not Kubelet?

Reviewer: yes

the device they manage. An access control must be set in place to restrict the
write access to the appropriate driver (A device status can only be updated by
the entities that have a direct control over the device(s) being reported).

Adding `Data` as an arbitrary data field may introduce extra processing
and storage overhead which might impact performance in a cluster with many
devices and frequent status updates. In large-scale clusters where many devices
are allocated, this impact must be considered.

## Design Details

### API

The `ResourceClaimStatus` struct in `pkg/apis/resource/types.go` will be
extended to include the slice of `Devices`.

`ResourceClaim` validation of the status in
`pkg/apis/resource/validation/validation.go` will be covered to allow a device
to be reported only once in the slice, a device is being identified  by
`<driver name>/<pool name>/<device name>`.

### Write Permission

To prevent unauthorized or accidental modifications by entities that do not 

> have access to a particular resource, a `ValidatingAdmissionPolicy` will be 

Reviewer 6: I do not love this trend - does it need a VAP per driver?

unique format, the update of the devices in the status will be rejected.

Here is a `ResourceClaim` allocated on a node. This would only work for now if 
exactly one node is set:
```yaml
apiVersion: resource.k8s.io/v1alpha3
kind: ResourceClaim
metadata:
  ...
spec:
  ...
status:
  allocation:
    ...
    nodeSelector:
      nodeSelectorTerms:
      - matchFields:
        - key: metadata.name
          operator: In
          values:
          - my-node
  ...
```

Here is an example of how the `ValidatingAdmissionPolicy` could look like:
```yaml
---
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingAdmissionPolicy
metadata:
  name: "resourceclaim-device-status-update"
spec:
  failurePolicy: Fail
  matchConstraints:
    resourceRules:
    - apiGroups: ["resource.k8s.io"]
      apiVersions: ["*"]
      operations: ["UPDATE"]
      resources: ["resourceclaims"]
  matchConditions:
    - name: 'device-status-update'
      expression: >- # Validation only for objects with their .status.devices updated.
        object.status.devices != oldObject.status.devices
  validations:
  - expression: >- # User node must be the same node as the one where the ResourceClaim is allocated.

>       variables.userNodeName != variables.objectNodeName

Reviewer 6: Aren't VAP's asserting the affirmative?  I.e. this should be `==`  ?  If not, I am confused about my own VAPs.

Author: I will re-try this VAP with my PoC,`validations` should work fine, but I remember having issues with the `matchConditions`

    Name string `json:"name"`

    // InterfaceName is the network interface name inside the Pod for this attachment.
    // Examples: eth1 or net1
    //
    // +optional
    InterfaceName string `json:"interfaceName"`

    // ip is an IP address (IPv4 or IPv6) assigned to the pod
    IP string `json:"ip,omitempty"`

    // IsDefaultGW is a flag indicating that the interface with this IP
    // inside the Pod holds the Default Gateway.
    //
    // +optional
    IsDefaultGW bool `json:"isDefaultGW,omitempty"`
}
```

### KEP-4680 extension

During the WG Device Management meeting on 17th of September 2024 ([Slack
summary](https://kubernetes.slack.com/archives/C0409NGC1TK/p1726679433650409)),
the idea was to extend the [KEP-4680 about resource health status in the
`Pod.Status`](https://github.com/kubernetes/enhancements/issues/4680) in order
to expose device information and not just the health.

### Custom Resources

In the `ResourceClaim.Status.Devices`, instead of having opaque field (`Data`) and
specific type fields, an object reference could be used for each device. The custom 
object would be created and maintained by the driver to report the status of the 

> devices.

Reviewer 8: I like this. It's very Gateway-API-ish (eg, [ParametersReference](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.ParametersReference))

Author: This would be my last choice, as if you want to build a controller (e.g. EndpointSlice controller) that gets the IPs for some pods, then you will need to find those pods, then get the ResouceClaims for those pods (current scenario with what is being proposed), then get the status object for those ResourceClaims.

The best would be the `New Pod.Status Field` alternative, but from what I understood in the [Multi-Network KEP](https://github.com/kubernetes/enhancements/pull/3700) and some other discussion (e.g. [here](https://kubernetes.slack.com/archives/C0409NGC1TK/p1726679433650409)), this is not possible.

Reviewer 8: @thockin CRXes!

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

- [x] Metrics

>   - Metric name: ["resourceclaim_status_devices_update_attempts_total", "resourceclaim_status_devices_update_failures_total"]

Reviewer: I meant on two lines!
